content/issue-2/applicative-parsing.md

```diff
@@ -0,0 +1,716 @@
+(Functor mapped with a function that changes nothing stays the same) and preserving composition
+of functions (functor transformed with two functions should give the same result as the same
+functor transformed with a composition of those two functions). For as long as it only changes
+the values in a context without affecting the context itself - implementation should be valid.
+
+At this point the parser can represent arguments of any type and provides a way to map between types.
+
+## Handling failures
+
+Supporting mapping that can't fail is easy with just the `Functor` abstraction, but it can't
+handle failures well - consider parsing numeric information:
+
+```
+$ app --jobs 42
+$ app --jobs Forty-two
+```
+
+A parser using `map`, `FromStr::from_str` and `unwrap` can parse the first line but panics
+with a bad error message for the second one. To handle parsing failures `Magic` needs to be
+able to represent them. The usual approach for dealing with failing computations in Rust is
+with help from the `Result` type.
+
+```rust
+struct Magic<T>(Result<T, String>);
+impl<T> Magic<T> {
+    fn run(self) -> Result<T, String> {
+        self.0
+    }
+
+    fn parse<R, F: Fn(T) -> Result<R, String>>(self, f: F) -> Magic<R> {
+        match self.0 {
+            Ok(x) => Magic(f(x)),
+            Err(err) => Magic(Err(err)),
+        }
+    }
+```
+
+The `parse` method applies a failing computation on a value inside `Magic` if one exists or keeps
+the error message untouched otherwise. The change in the representation of `Magic` also
+requires to change `map`, but after that, fallible conversions are now possible:
+
+```rust
+impl<T> Magic<T> {
+    fn map<R, F: Fn(T) -> R>(self, f: F) -> Magic<R> {
+        self.parse(|x| Ok(f(x)))
+    }
+}
+
+#[test]
+fn sample() {
+    use std::str::FromStr;
+    // --jobs 42
+    let jobs = Magic(Ok("42"));
+    let result = answer
+        .parse(|n| u32::from_str(n).map_err(|e| e.to_string()))
+        .run();
+    assert_eq!(result, Ok(42));
+
+    let jobs = Magic(Ok("Forty-two"));
+    let result = answer
+        .parse(|n| u32::from_str(n).map_err(|e| e.to_string()))
+        .run();
+    assert_eq!(result, Err("invalid digit found in string".to_owned()))
+}
+```
+
+At this point the parser can represent arguments of any type and failures too.
+Alas, `parse` is an ad-hoc thing and isn't coming from Category Theory.
+
+## Composing failing computations
+
+`run` makes it easy to check if computation is successful and to proceed only when it is, but
+it also adds a new corner case: consider an app that takes a name and the answer, checks if the
+answer is correct and reports to the user:
+
+```rust
+let bin = match bin.run().expect("You need to specify --bin");
+
+// println!("Building binary {bin}!") // (1)
+
+let jobs = match jobs.run().expect("You need to specify --jobs");
+
+// println!("Building binary {bin}!") // (1)
```

Comment:
This line should be (2)?

Reply:
Right, that was me doing copy-paste when renaming things.

```suggestion
// println!("Building binary {bin}!") // (2)
```